### ISSUE TITLE
Inclusion of PRIMA solvers in Skit-Optimize

### ISSUE THREAD

**Issue by zaikunzhang** [scikit-optimize/scikit-optimize#1182]
Dear Scikit-Optimize maintainers,

This is Dr. Zaikun Zhang from The Hong Kong Polytechnic University. Together with Professor [N.I.M. Gould](https://www.numerical.rl.ac.uk/people/nimg/), I am responsible for maintaining the renowned derivative-free optimization solvers of the late Professor [M.J.D. Powell](https://www.zhangzk.net/powell.html), namely COBYLA, UOBYQA, NEWUOA, BOBYQA, and LINCOA. I am the author of [PRIMA](http://www.libprima.net/), which provides the reference implementation for these solvers. They are widely used by engineers and scientists. For instance, see Section 1 of [a recent paper on Powell's solvers](https://arxiv.org/pdf/2302.13246.pdf) as well as the Google searches of [COBYLA](https://www.google.com/search?q=cobyla) and [BOBYQA](https://www.google.com/search?q=bobyqa).

Since your package is exactly oriented to derivative-free / black-box optimization, it might be desirable to include the PRIMA solvers. I will be happy to assist on the Fortran side if you would like to do so.

Note that, even though the old Fortran 77 implementation of the aforementioned solvers is truly a masterpiece, it contains many bugs (mostly due to the language itself), which can lead to segmentation faults or infinite loops. For example, see [Section 4.4 of the above paper](https://arxiv.org/pdf/2302.13246.pdf) and [many GitHub issues](https://github.com/libprima/prima#bug-fixes). It is strongly discouraged to use the Fortran 77 version of these solvers anymore.

Thanks and regards,
Zaikun ZHANG
Ph.D. and Assistant Professor
Dept. App. Math., Hong Kong Polytechnic University

**Comment by Deathn0t** [comment]
Dear @zaikunzhang , 

As a user of this package I think the `scikit-optimize` original repository is not maintained anymore. However, I am working on an updated fork of this package available [here]( https://github.com/deephyper/deephyper/tree/master/deephyper/skopt) within an hyperparameter optimization software called [DeepHyper](https://github.com/deephyper/deephyper/tree/master/deephyper). Also, I was wondering if your approach could be used to **optimize the acquisition function** within Bayesian optimization. In replacement of `lbfgs` in scikit-optimize for example (would it perform better?).

Let me know if you are interested to try this out. I have implemented genetic algorithm recently to compare against lbfgs.
Best regards,
Romain EGELE,
Ph.D. at Universite Paris-Saclay.

**Comment by zaikunzhang** [comment]
Hi Romain @Deathn0t ,

The solvers in PRIMA can solve black-box optimization problems without requiring derivatives (first-order information). They are quite capable in solving hyper-parameter tuning problems. So they are good solvers to be included in DeepHyper. 

>  if your approach could be used to optimize the acquisition function within Bayesian optimization. In replacement of lbfgs in scikit-optimize for example (would it perform better?).

Sure. Whether it will outperform lbfgs or not, it depends on the properties of the problem. If your problem is a black-box optimization problem, the PRIMA solvers are the state of the art --- they definitely outperform genetic algorithms. However, it your problem can provide first-order information, then gradient-based algorithms will perform better. 

Thanks,
Zaikun

**Comment by Deathn0t** [comment]
Thanks for your reply @zaikunzhang . Also, I have the following questions:

1. Is it better to wait for https://github.com/scipy/scipy/issues/18118 and https://github.com/libprima/prima/pull/52 to be completed to use PRIMA with Python or is there a Python binding already usable?
2. Does it work with mixed-variable spaces (Real, Discrete, Categorical)?
3. Is there an "asynchronous" interface of the type `ask/tell` or `suggest/observe` to use prima in an asynchronous parallel setting?
4. Can it perform batch vectorized optimization? (instead of inferring the function once per input point, we infer a list of points directly, it can be faster in Python to avoid slow Python for loops)

Thanks,
Romain.

**Comment by zaikunzhang** [comment]
Hi,

> Is it better to wait for https://github.com/scipy/scipy/issues/18118 and https://github.com/libprima/prima/pull/52 to be completed to use PRIMA with Python or is there a Python binding already usable?

Yes, you can wait for  https://github.com/libprima/prima/pull/52 . 

> Does it work with mixed-variable spaces (Real, Discrete, Categorical)?

No.

> Is there an "asynchronous" interface of the type ask/tell or suggest/observe to use prima in an asynchronous parallel setting?

Not directly. Maybe you can implement that using a callback function, but I am not sure. 

> Can it perform batch vectorized optimization? (instead of inferring the function once per input point, we infer a list of points directly, it can be faster in Python to avoid slow Python for loops)

Since PRIMA solvers are intrinsically sequential (except for the initialization), I guess the answer is no.  However, I do not think calling PRIMA will involve Python for loops --- it will be a single line of Python code, behind which is the Fortran code. 

Thank you. 

Zaikun

**Comment by Deathn0t** [comment]
Thank you for the detailed replies @zaikunzhang !
